### PR TITLE
Host changes for azure-functions-cli

### PR DESCRIPTION
**Pull request by ahmelsayed**

### CODE REVIEW

File: src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
         public string LogPath { get; set; }
+
+        [JsonProperty("secretsPath")]
         public string SecretsPath { get; set; }
+
+        [JsonProperty("nodeDebugPort")]
+        public int NodeDebugPort { get; set; }
+
+        [JsonIgnore]
+        public TraceWriter TraceWriter { get; set; }
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs]: This is causing an issue when picked up by the `XmlMediaTypeFormatter` since it relies on the `DataContractSerializer` and doesn't respect this attribute.

File: src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs [hunk 2]
```diff
@@ -23,6 +25,8 @@ internal static void Initialize(ScriptSettingsManager settingsManager, Container
             builder.Register<ISecretManager>(ct => ct.Resolve<WebHostResolver>().GetSecretManager(settings)).ExternallyOwned();
             builder.Register<WebScriptHostManager>(ct => ct.Resolve<WebHostResolver>().GetWebScriptHostManager(settings)).ExternallyOwned();
             builder.Register<WebHookReceiverManager>(ct => ct.Resolve<WebHostResolver>().GetWebHookReceiverManager(settings)).ExternallyOwned();
+            builder.RegisterInstance(settings);
+            builder.RegisterHttpRequestMessage(config);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs]: Is this needed? What does it do?

File: src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public static class WebApiConfig
 
             var builder = new ContainerBuilder();
             builder.RegisterApiControllers(typeof(FunctionsController).Assembly);
-            AutofacBootstrap.Initialize(settingsManager, builder, settings);
+            AutofacBootstrap.Initialize(settingsManager, builder, settings, config);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs]: might not be needed based on question above

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 4]
```diff
@@ -87,7 +93,10 @@ public HostStatus GetHostStatus()
         {
             HostStatus status = new HostStatus
             {
-                Id = _scriptHostManager.Instance?.ScriptConfig.HostConfig.HostId
+                Id = _scriptHostManager.Instance?.ScriptConfig.HostConfig.HostId,
+                WebHostSettings = _webHostSettings,
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Thinking we should only return WebHostSettings when running locally - not really useful otherwise I think?

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 5]
```diff
@@ -87,7 +93,10 @@ public HostStatus GetHostStatus()
...
+                ProcessId = Process.GetCurrentProcess().Id,
+                IsDebuggerAttached = Debugger.IsAttached
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: This will return the correct value for Antares remote debugging as well?

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 6]
```diff
@@ -59,8 +62,12 @@ public HttpResponseMessage Invoke(string name, [FromBody] FunctionInvocation inv
         [Route("admin/functions/{name}/status")]
         public FunctionStatus GetFunctionStatus(string name)
         {
-            FunctionStatus status = new FunctionStatus();
             Collection<string> functionErrors = null;
+            FunctionDescriptor function = _scriptHostManager.Instance.Functions.FirstOrDefault(p => p.Name.ToLowerInvariant() == name.ToLowerInvariant());
+            FunctionStatus status = new FunctionStatus
+            {
+                Metadata = function?.Metadata
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: What bits of the metadata do you actually need for CLI? Wondering if instead of shoving this into this status api, we shouldn't instead have a new admin/functions/{name}/metadata api? This API is really about returning status only, and we shouldn't return any irrelevant info I think.

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 7]
```diff
@@ -100,6 +109,17 @@ public HostStatus GetHostStatus()
             return status;
         }
 
+        [HttpPost]
+        [Route("admin/host/debug")]
+        public bool LaunchDebugger()
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Can we return an HttpResponseMessage instead here, so we can return 200 if successfull, or if not local, an appropriate error response?

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 8]
```diff
@@ -100,6 +109,17 @@ public HostStatus GetHostStatus()
...
+        {
+            if (_webHostSettings.IsSelfHost)
+            {
+                return Debugger.Launch();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Is this idempotent? E.g. what happens if you call this 5 times in a row? Should you use Debugger.IsAttached as a guard, or is it ok to call repeatedly?

File: src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs (hunk 1 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs]: Also, I'm thinking that for the purposes of returning a DTO from the admin api, it might be good to just define a new type that defines what you want to return. This type is kind of an internal settings class that may grow over time, and I don't want that to result in us returning more than we intended inadvertently.

File: src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs [hunk 9]
```diff
@@ -93,6 +93,7 @@ private static WebHostSettings GetDefaultSettings(ScriptSettingsManager settings
                 settings.ScriptPath = settingsManager.GetSetting(EnvironmentSettingNames.AzureWebJobsScriptRoot);
                 settings.LogPath = Path.Combine(Path.GetTempPath(), @"Functions");
                 settings.SecretsPath = HttpContext.Current.Server.MapPath("~/App_Data/Secrets");
+                settings.IsSelfHost = true;
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs]: Do all the e2e tests pass with this set? Other of your changes below force an auth level of Admin when self hosted. I'd expect that to cause our e2e auth tests to fail.

Comment (fabiocav) [inline src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs]: This overloads `IsSelfHost` which is currently used to indicate when self hosting in a console app. Local runs in IIS/IIS Express should continue to have this return `false`

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs [hunk 10]
```diff
@@ -137,19 +138,27 @@ private static void ReinitializeAppSettings()
             }
         }
 
-        private static ScriptHostConfiguration GetScriptHostConfiguration(string scriptPath, string logPath)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: Can you explain this change? Looks like we had an overload previously that was only used by tests, is that is?

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs [hunk 11]
```diff
@@ -168,6 +177,51 @@ private static ScriptHostConfiguration GetScriptHostConfiguration(string scriptP
             return scriptHostConfig;
         }
 
+        private static string MakeValidHostId(string id)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: I think all this logic to default the host ID to the local machine name should go in the CLI. I believe there is already logic in the CLI to default the host ID to machine name?

We've been running fine up to this point locally - the guidance is for people to simply put an ID in their host.config, as we do in the sample app.

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 6 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: I look at ScriptType to determine the language for debugging, and InputBinding for `IsHttpFunction` check when running.

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 5 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: correct

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 4 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: sure, though just out of curiosity, why?

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 8 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: it does that internally (https://referencesource.microsoft.com/#mscorlib/system/diagnostics/debugger.cs,94)

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs (hunk 11 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: Also, you should consider the implications of deriving an ID from the current directory path. The ID for a function app (set of functions) is supposed to be a static identifier that doesn't change. Logging and other storage artifacts depend on that. E.g. when set explicitly in host.json that ID will travel around with the files. In your case it will change if the app is copied to a different directory, machine, etc.

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 12]
```diff
@@ -41,11 +41,12 @@
     <DocumentationFile>
     </DocumentationFile>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(OS)'!='Unix'">
+    <EnableNode>True</EnableNode>
+    <EnablePowerShell>True</EnablePowerShell>
+    <DefineConstants>$(DefineConstants);WINDOWS;FEATURE_NODE;FEATURE_POWERSHELL</DefineConstants>
```
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: I'm hoping we can simplify all these various switches (MONO,WINDOWS,FEATURE_NODE,FEATURE_POWERSHELL) to just a single MONO switch. We're not going to actually having all these various permutations right? We can just simplify all these conditionals in the code to use just a single MONO constant.

File: src/WebJobs.Script/PlatformHelper.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    internal static class PlatformHelper
+    {
+        public static bool IsMono { get; } = Type.GetType("Mono.Runtime") != null;
+        public static bool IsWindows { get; } = Environment.OSVersion.Platform == PlatformID.Win32NT;
```
Comment (mathewc) [inline src/WebJobs.Script/PlatformHelper.cs]: I don't see IsWindows being used anywhere. Let's just reduce this to a single flag.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 14]
```diff
@@ -315,12 +317,16 @@ protected virtual void Initialize()
                 List<FunctionDescriptorProvider> descriptionProviders = new List<FunctionDescriptorProvider>()
                 {
                     new ScriptFunctionDescriptorProvider(this, ScriptConfig),
+#if FEATURE_NODE
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: If you compress all these flags to one (see comments below), then we can simplify this code to just a single `#if !MONO` block

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 15]
```diff
@@ -405,14 +411,16 @@ private void RestartHost()
...
             // whenever we're restarting the host, we want to let the Node
             // invoker know so it can clear the require cache, etc.
             NodeFunctionInvoker.OnHostRestart();
+#endif
         }
 
-        /// <summary>
-        /// Whenever the debug marker file changes we update our debug timeout
-        /// </summary>
+            /// <summary>
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: nit: revert these formatting changes?

File: src/WebJobs.Script.WebHost/Properties/Resources.resx [hunk 16]
```diff
@@ -183,7 +183,7 @@
   </data>
   <assembly alias="System.Windows.Forms" name="System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
   <data name="Homepage" type="System.Resources.ResXFileRef, System.Windows.Forms">
-    <value>..\Home.html;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8</value>
+    <value>../Home.html;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8</value>
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Properties/Resources.resx]: Is this change needed?

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs [hunk 17]
```diff
@@ -142,7 +142,12 @@ public void Initialize()
                     {
                         if (!_webHostSettings.IsSelfHost)
                         {
+                            // FIXME: Mono does not yet support HostingEnvironment.QueueBackgroundWorkItem
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: nit: can you change this to a TODO? That's what we use in this codebase :)

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 18]
```diff
@@ -41,11 +41,12 @@
...
+  </PropertyGroup>
   <ItemGroup>
-    <Reference Include="EdgeJs, Version=0.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Rather than completely restructure/reorder these project files into item groups, don't the Reference/Compile elements support a Condition attribute so this can simply be tacked on as needed? This will make this much easier to review and maintain. E.g. in this review I'd then be able to see that you're simply adding Conditions to existing elements - with the reordering I honestly can't tell if you've accidentally added/removed things, etc.

File: test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs [hunk 19]
```diff
@@ -32,7 +32,7 @@ public ScriptHostManagerTests()
         public async Task UpdateFileAndRestart()
         {
             CancellationTokenSource cts = new CancellationTokenSource();
-            var fixture = new NodeEndToEndTests.TestFixture();
+            var fixture = new CSharpEndToEndTests.TestFixture();
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs]: Can this change be reverted?

File: tools/NuGetProj.targets [hunk 20]
```diff
@@ -92,6 +92,11 @@ Common build targets overwrites.
     <CleanDependsOn>BeforeClean;ReadNuGetCleanOutputs;CoreClean;AfterClean</CleanDependsOn>
     <RebuildDependsOn>Clean;Build</RebuildDependsOn>
   </PropertyGroup>
+  <!-- FIXME disable nuget pack on Unix as it currently fails -->
```
Comment (mathewc) [inline tools/NuGetProj.targets]: Rather than make this change in this tools file, can you make it in the WebJobs.Script.proj to exclude the nuget packaging?

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 6 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: so the problem with getting that from `WebScriptHostManager` is that we can't new up that object. The only way we know in the cli how to get a hold of that object is by actually instantiating the host itself, then grab it from the dependency resolver we get back.

File: src/WebJobs.Script/PlatformHelper.cs (hunk 13 above)
Comment (mhutch) [inline src/WebJobs.Script/PlatformHelper.cs]: IsWindows was used by the CLI code, which was split out. Should be fine to remove it from this copy of the file.

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 12 above)
Comment (mhutch) [inline src/WebJobs.Script/WebJobs.Script.csproj]: I did the various switched to keep the semantics of the individual changes clearer and make it easier to add node/powershell support on unix in future.

i.e.

* `MONO`: changes needed to run on the mono _runtime_ (which can run on windows).
* `WINDOWS`: things that will only work on the windows _platform_.
* `FEATURE_NODE`: the Node feature, which is not yet implemented for Unix (but could be)
* etc.

I imagine these annotations might be useful for the .NET Core port?

File: src/WebJobs.Script.WebHost/Properties/Resources.resx (hunk 16 above)
Comment (mhutch) [inline src/WebJobs.Script.WebHost/Properties/Resources.resx]: Yes. The Mono resx compiler doesn't transparently remap backslashes, and IIRC the Windows one can deal with forward slashes.

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs (hunk 17 above)
Comment (mhutch) [inline src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: Ah, we use both. The distinction is subtle ;)

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 18 above)
Comment (mhutch) [inline src/WebJobs.Script/WebJobs.Script.csproj]: They do, I just thought this way was neater.

File: test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs (hunk 19 above)
Comment (mhutch) [inline test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs]: Not trivially, because `NodeEndToEndTests` is omitted when building for Mono. I could `#if` out `UpdateFileAndRestart`, I guess.

File: tools/NuGetProj.targets (hunk 20 above)
Comment (mhutch) [inline tools/NuGetProj.targets]: I tried that and didn't see an easy way to do it.

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs (hunk 11 above)
Comment (mhutch) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: FWIW, this logic was already there, I just factored it into a separate method and made it more robust.

**Comment by mhutch** [comment]
I had to fix up the patch to get it working on Mono again: https://github.com/mhutch/azure-webjobs-sdk-script/commit/4605d0e21b636e8a9e1b692e1b14524f34c9e0f7

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs (hunk 11 above)
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: This was logic added for the CLI changes which is where you (correctly) based your work, which is why is showing as an addition to the runtime. We should be able to use your implementation when this logic is moved to the CLI (depending on how @ahmelsayed chooses to approach it there)

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs (hunk 10 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: I believe actually most of these changes here were added by @fabiocav if I'm not mistaken.
https://github.com/Azure/azure-webjobs-sdk-script/commit/6b767e1022c9f31b2be0e2a842fceb0084c934c4

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 7 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: does this make sense to you 

```csharp
if (_webHostSettings.IsSelfHost)
{
    if (Debugger.Launch())
    {
        return new HttpResponseMessage(HttpStatusCode.Created);
    }
    else
    {
        return new HttpResponseMessage(HttpStatusCode.OK);
    }
}
return new HttpResponseMessage(HttpStatusCode.NotImplemented);
```

**Comment by ahmelsayed** [comment]
@mathewc and @fabiocav I believe I addressed all the feedback except for the inclusion of `Metadata` in the function status API. I explained [here](https://github.com/Azure/azure-webjobs-sdk-script/pull/1083#discussion_r95047218) why it's needed at the moment. How do you wanna proceed with that?

**Comment by ahmelsayed** [comment]
@mathewc, @fabiocav, regarding the auth test cases, how do you want me to fix that?
Basically this [line ](https://github.com/Azure/azure-webjobs-sdk-script/blob/split-cli/src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs#L47) disables all auth when `IsSelfHost` is true. This is breaking the auth tests.

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs (hunk 11 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: Ok, so we're moving this out?

Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: Oh sorry forgot to do that. Removed!

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 6 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: I've made a simple refactoring to expose a new public ScriptHost.ReadFunctionMetadata function that you can call (see PR and example in test here https://github.com/Azure/azure-webjobs-sdk-script/pull/1117). This was easy and is preferable I think than adding new http surface area for this.

File: src/WebJobs.Script.WebHost/Properties/Resources.resx (hunk 16 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Properties/Resources.resx]: We just need to be sure this isn't breaking for Windows

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 12 above)
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Let's simplify all this to just a single MONO flag. If we need additional complexity in the future for some reason we can add it later. That satisfies the requirements of this very temporary situation where we need to do compilation for Mac/Mono. When we move to .NET core this all goes away.

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 6 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Sounds good. Updated the PR.

File: src/WebJobs.Script.WebHost/Properties/Resources.resx (hunk 16 above)
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Properties/Resources.resx]: There's a test that validates the result when hitting the home controller (which uses that resource) and that seems to be passing, so looks like it is happy with the forward slash on Windows.

File: src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs (hunk 9 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs]: @ahmelsayed If you really want to disable all auth for CLI, then you should add a new flag (with big warnings on it) WebHostSettings.DisableAuth.

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 21]
```diff
@@ -100,6 +107,25 @@ public HostStatus GetHostStatus()
...
+        [HttpPost]
+        [Route("admin/host/debug")]
+        public HttpResponseMessage LaunchDebugger()
+        {
+            if (_webHostSettings.IsSelfHost)
+            {
+                // If debugger is already running, this will be a no-op returning true.
+                if (Debugger.Launch())
+                {
+                    return new HttpResponseMessage(HttpStatusCode.Created);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: nit: Accepted might be more correct

File: src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs [hunk 22]
```diff
@@ -1,13 +1,28 @@
...
 
+using System.Runtime.Serialization;
+using Microsoft.Azure.WebJobs.Host;
+using Newtonsoft.Json;
+
 namespace Microsoft.Azure.WebJobs.Script.WebHost
 {
     public class WebHostSettings
     {
+        [JsonProperty("isSelfHost")]
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs]: I think you can just use the existing HostStatus.Id property for your purposes and revert all these changes in this file.

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 23]
```diff
@@ -87,7 +90,11 @@ public HostStatus GetHostStatus()
         {
             HostStatus status = new HostStatus
             {
-                Id = _scriptHostManager.Instance?.ScriptConfig.HostConfig.HostId
+                Id = _scriptHostManager.Instance?.ScriptConfig.HostConfig.HostId,
+                // Only return WebHostSettings if running locally as it's not useful otherwise.
+                WebHostSettings = _webHostSettings?.IsSelfHost == true ? _webHostSettings : null,
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Can remove this once you're using HostId instead?

File: src/WebJobs.Script.WebHost/Models/FunctionStatus.cs [hunk 24]
```diff
@@ -13,5 +14,8 @@ public class FunctionStatus
         /// </summary>
         [JsonProperty(PropertyName = "errors", DefaultValueHandling = DefaultValueHandling.Ignore)]
         public Collection<string> Errors { get; set; }
+
+        [JsonProperty(PropertyName = "metadata", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public FunctionMetadata Metadata { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Models/FunctionStatus.cs]: You can remove this now?

File: src/WebJobs.Script.WebHost/Models/HostStatus.cs [hunk 25]
```diff
@@ -30,5 +31,29 @@ public HostStatus()
         /// </summary>
         [JsonProperty(PropertyName = "errors", DefaultValueHandling = DefaultValueHandling.Ignore)]
         public Collection<string> Errors { get; set; }
+
+        /// <summary>
+        /// Gets or sets the web host settings.
+        /// </summary>
+        [JsonProperty(PropertyName = "webHostSettings", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public WebHostSettings WebHostSettings { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Models/HostStatus.cs]: Remove

File: src/WebJobs.Script.WebHost/Models/HostStatus.cs [hunk 26]
```diff
@@ -30,5 +31,29 @@ public HostStatus()
...
+
+        /// <summary>
+        /// Gets or sets the processId for the host.
+        /// </summary>
+        [JsonProperty(PropertyName = "processId", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public int ProcessId { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Models/HostStatus.cs]: Can you remove this?

File: src/WebJobs.Script.WebHost/Models/HostStatus.cs [hunk 27]
```diff
@@ -30,5 +31,29 @@ public HostStatus()
...
+
+        /// <summary>
+        /// Gets or sets whether the process has a debugger attached or not.
+        /// </summary>
+        [JsonProperty(PropertyName = "isDebuggerAttached", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public bool IsDebuggerAttached { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Models/HostStatus.cs]: Do you still need this?

File: src/WebJobs.Script.WebHost/Models/HostStatus.cs [hunk 28]
```diff
@@ -30,5 +31,29 @@ public HostStatus()
...
+
+        internal static string GetAssemblyFileVersion(Assembly assembly)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Models/HostStatus.cs]: Move back to ScriptHost?

Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Models/HostStatus.cs]: I think this was a merge mistake. You moved it from here back in August https://github.com/Azure/azure-webjobs-sdk-script/commit/0c50aa890f7a5a81aeb7c24676ec47647e7efc92.
Removed.

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 18 above)
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Ok, we can leave this as is.

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 12 above)
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: We'll probably compress all these down ourselves after this gets checked in.

File: src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs (hunk 9 above)
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs]: added https://github.com/Azure/azure-webjobs-sdk-script/pull/1083/commits/638a0a1f928f3deb04cbc3125c8a9ec6be9602dc#diff-12f35aef698cf86bad7b8cea0e786225R19

File: test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs (hunk 19 above)
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs]: @ahmelsayed Please revert the changes to this file before checking in - these tests are intentionally using a Node host. If needed you can just conditional out the file based on the MONO flag in the test proj file.

File: src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs (hunk 9 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebApiConfig.cs]: Thanks. Can you also remove this settings.IsSelfHost = true line now?

File: src/WebJobs.Script.WebHost/Filters/AuthorizationLevelAttribute.cs [hunk 29]
```diff
@@ -32,8 +32,9 @@ public AuthorizationLevelAttribute(AuthorizationLevel level)
             }
 
             ISecretManager secretManager = actionContext.ControllerContext.Configuration.DependencyResolver.GetService<ISecretManager>();
+            var settings = actionContext.ControllerContext.Configuration.DependencyResolver.GetService<WebHostSettings>();
 
-            if (!await IsAuthorizedAsync(actionContext.Request, Level, secretManager))
+            if (!settings.IsSelfHost && !await IsAuthorizedAsync(actionContext.Request, Level, secretManager))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Filters/AuthorizationLevelAttribute.cs]: You mean to use your new DisableAuth setting here. Also let's rename it to IsAuthDisabled to match the Is pattern of the other bools.

Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Filters/AuthorizationLevelAttribute.cs]: Right, sorry I missed that.
